### ISSUE TITLE
Should `:Neotree source=example` work?

### ISSUE THREAD

**Issue by ElPiloto** [nvim-neo-tree/neo-tree.nvim#396]
I'm trying to implement my own custom source and thought I could base it off of the `sources/example` source.  However, `:Neotree source=example` gives me the error below.  I'm literally just getting started with this codebase today so it's likely I'm doing something wrong.  Should I not try to use the example source?

```
E5108: Error executing lua ...ker/start/neo-tree.nvim/lua/neo-tree/sources/manager.lua:39: attempt to index local 'state' (a nil value)
stack traceback:
	...ker/start/neo-tree.nvim/lua/neo-tree/sources/manager.lua:39: in function 'create_state'
	...ker/start/neo-tree.nvim/lua/neo-tree/sources/manager.lua:111: in function 'get_state'
	...packer/start/neo-tree.nvim/lua/neo-tree/command/init.lua:79: in function 'execute'
	...packer/start/neo-tree.nvim/lua/neo-tree/command/init.lua:140: in function '_command'
	[string ":lua"]:1: in main chunk
```

Edit to add some further context: I did add "example" to the source list in `setup/init.lua` and a dummy config for example in `defaults.lua`

**Comment by cseickel** [comment]
It doesn't work, sorry about that! Looks like I haven't maintained it through some breaking changes. I'll update it.

You may want to start with the `buffers` source as the example. It's pretty simple and it actually works.

One other thing you'll need to do when creating a new source is to add it to this table: 
https://github.com/nvim-neo-tree/neo-tree.nvim/blob/8613c458cd04ddd14135060179bc2bca01e2d2f9/lua/neo-tree/setup/init.lua#L14

BTW - What source are you going to add?

**Comment by ElPiloto** [comment]
Actually, it's still a relevant question. I'm planning on creating:
- bookmarks source (likely piggybacking on ezbookmarks.nvim)
- a "workspace" source, I put that in quotes because it's actually something specific to internal VCS tooling we use at my job

I did manage to get something working for the "workspace" source (no actions yet though) by copying the filesystem source.

I'm still trying to grok all the parts of the codebase though (e.g. rendering), so I'm sure I'd benefit from a fix to the example source or any other documentation on how to add a source.

**Comment by cseickel** [comment]
> I'm still trying to grok all the parts of the codebase though (e.g. rendering),

The good news is that you do not really need to grok rendering to add a new source. Treat it as a black box that just works. Everything you need to add is within the sources folder. If you want to customize the display of something, add a new component.

**Comment by cseickel** [comment]
The example source was fixed in 4e6592670dcac51cc2d0ea97e9e33792ac5b132d, which is in `main` right now.

When you have questions about how to do things, head on over to the [Discussion](https://github.com/nvim-neo-tree/neo-tree.nvim/discussions/categories/internal-development) section and start a conversation in the **Internal Development** category.
